request.go

```diff
@@ -0,0 +1,251 @@
+	}
+
+	httpReq, err := http.NewRequest(r.fields.Method, urlStr, content)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.fields.Content != nil {
+		httpReq.Header.Set("Content-Type", "application/json")
+	}
+
+	for keyID, sig := range r.fields.Signatures[r.fields.Origin] {
+		httpReq.Header.Add("Authorization", fmt.Sprintf(
+			"X-Matrix origin=\"%s\",key=\"%s\",sig=\"%s\"", r.fields.Origin, keyID, sig,
+		))
+	}
+
+	return httpReq, nil
+}
+
+// VerifyHTTPRequest extracts and verifies the contents of a net/http.Request.
+// It consumes the body of the request.
+// The JSON content can be accessed using MatrixRequest.Content()
+// Returns an 400 error if there was a problem parsing the request.
+// It authenticates the request using an ed25519 signature using the KeyRing.
+// The origin server can be accessed using MatrixRequest.Origin()
+// Returns a 401 error if there was a problem authenticating the request.
+// HTTP handlers using this should be careful that they only use the parts of
+// the request that have been authenticated: the method, the request path,
+// the query parameters, and the JSON content.
+func VerifyHTTPRequest(
+	req *http.Request, now time.Time, destination string, keys KeyRing,
+) (*MatrixRequest, util.JSONResponse) {
+	request, err := readHTTPRequest(req)
+	if err != nil {
+		util.GetLogger(req.Context()).WithError(err).Print("Error parsing HTTP headers")
+		return nil, util.MessageResponse(400, "Bad Request")
+	}
+	request.fields.Destination = destination
+
+	// The request fields are already in the form required by the specification
+	// So we can just serialise the request fields using the default marshaller
+	toVerify, err := json.Marshal(request.fields)
+	if err != nil {
+		util.GetLogger(req.Context()).WithError(err).Print("Error parsing JSON")
+		return nil, util.MessageResponse(400, "Invalid JSON")
+	}
+
+	if request.Origin() == "" {
+		message := "Missing \"Authorization: X-Matrix ...\" HTTP header"
+		util.GetLogger(req.Context()).WithError(err).Print(message)
+		return nil, util.MessageResponse(401, message)
+	}
+
+	results, err := keys.VerifyJSONs([]VerifyJSONRequest{{
+		ServerName: request.Origin(),
+		AtTS:       Timestamp(now.UnixNano() / 1000000),
+		Message:    toVerify,
+	}})
+	if err != nil {
+		message := "Error authenticating request"
+		util.GetLogger(req.Context()).WithError(err).Print(message)
+		return nil, util.MessageResponse(500, message)
+	}
+	if results[0].Result != nil {
+		message := "Invalid request signature"
+		util.GetLogger(req.Context()).WithError(results[0].Result).Print(message)
+		return nil, util.MessageResponse(401, message)
+	}
+
+	return request, util.JSONResponse{Code: 200, JSON: struct{}{}}
+}
+
+// Returns an error if there was a problem reading the content of the request
+func readHTTPRequest(req *http.Request) (*MatrixRequest, error) {
+	var result MatrixRequest
+
+	result.fields.Method = req.Method
+	result.fields.RequestURI = req.URL.RequestURI()
```

Comment:
Do we need to worry about the way Go escapes/doesn't escape things such that this may not match the thing that was signed?